numpy/core/src/multiarray/number.c

```diff
@@ -176,6 +184,19 @@ PyArray_GetNumericOps(void)
     return NULL;
 }
 
+/*NUMPY_API
+  Get dictionary showing number functions that all arrays will use
+*/
+NPY_NO_EXPORT PyObject *
+PyArray_GetNumericOps(void)
```

Comment:
Should be wrapped in
```
#if !defined(NPY_NO_DEPRECATED_API) || \
    (NPY_NO_DEPRECATED_API < NPY_1_16_API_VERSION)

#endif
```

Reply:
That is only for headers, to allow access to old API macros. The C files always start with this define, which will defeat the check you suggest
[`#define NPY_NO_DEPRECATED_API NPY_API_VERSION`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/src/multiarray/number.c#L6). `NPY_API_VERSION` is hard-coded from `C_API_VERSION` in [`core/setup_common.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/setup_common.py#L44) , see [`core/setup.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/setup.py#L561) where the magic happens.

We don't really have a way to actually remove or disable deprecated API functions, since the convention is that the PyArray_API structure built in `__multiarray_api.h` by [`core/code_generators/generate_array_api.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/code_generators/generate_numpy_api.py) from the list in [`core.code_generators/numpy_api.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/code_generators/numpy_api.py#L32) can only grow, never shrink, to allow backward compatibility.

This is somewhat talked about in the [documentation](http://www.numpy.org/devdocs/reference/c-api.array.html?highlight=pyarray_api#miscellaneous) from the user's perspective and in the ['VERSIONING SUPPORT'](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/setup_common.py#L12) docstring in `setup_common.py`, but if you think of hook where I can hang more documentation let me know.

Reply:
We could eventually turn the warning into an exception. Not sure if I understand which "C api macro" you mean, there are so many.

At some point we could remove things from the `PyArray_API` structure by changing our major version, which would infer no backward API compatibility.

Reply:
I mean this macro:
```
#define PyArray_GetNumericOps \
        (*(PyObject * (*)(void)) \
         PyArray_API[41])
```

Perhaps we should leave this as an open question, and just take this PR as it currently stands